services/web/src/utils/middleware/l10n.js

```diff
@@ -0,0 +1,134 @@
+// Handles localization via a cookie or URL prefix.
+//
+// For default URLs with no prefix, this middleware
+// will check the "hl" cookie or the browser's
+// Accept-Language header. If a language matches
+// that is not the default language, then they will
+// be redirected to the page with the appropriate
+// prefix.
+//
+// For URLs with a prefix, if the prefix is the
+// default language then they will be redirected
+// to the canonical page without it, otherwise
+// they will be allowed on.
+//
+// This middleware will also set an "l10n" object on
+// the state containing contextual information about
+// the current language as well as canonical URLs for
+// the current page for use with meta tags.
+//
+// Note that this middleware intentionally does not
+// concern itself with actually setting the cookie
+// but instead allows this to be handled by either
+// Javascript or another middleware (via a param).
+//
+// Usage:
+//
+// app.use(l10n(['en-US', 'ja-JP'], 'en-US'))
+//
+// Examples:
+//
+//           Request                 |     Response
+// ----------------------------------|-------------------
+// URL       Cookie  Accept-Language |  URL       Status
+// ----------------------------------|-------------------
+// /page     none    en,ja;q=0.9     |  /page     200
+// /page     none    ja,en;q=0.9     |  /ja/page  301
+// /page     ja      en,ja;q=0.9     |  /ja/page  301
+// /en/page  none    en,ja;q=0.9     |  /page     301
+// /ja/page  none    en,ja;q=0.9     |  /ja/page  200
+// /ja/page  none    ja,en;q=0.9     |  /ja/page  200
+// /ja/page  en      en,ja;q=0.9     |  /ja/page  200
+//
+//
+
+const COOKIE_NAME = 'hl';
+const DEFAULT_CODE = 'en-US';
+
+module.exports = function l10nMiddleware(localeCodes, defaultCode) {
```

Comment:
also i doubt this is something that will be used

Reply:
You're probably right but it's something I spent a bit of time on and should be solid. If we ever need it I'd like it not to be lost to the mists of time. I just added a TODO to move it to a plugin when we get that facility... does that work?